lib/socket.rb

```diff
@@ -263,9 +263,30 @@ def protocol
     # TODO
   end
 
-  def ip_address
-    # TODO
-  end
+  __define_method__ :ip_address, <<-CPP
+      auto addrinfo = self->ivar_get(env, "@_addrinfo"_s);
+      if(addrinfo->is_nil())
+          env->raise("SocketError", "need IPv4 or IPv6 address");
+
+      auto info_struct = (struct addrinfo *)addrinfo->as_void_p()->void_ptr();
+
+      switch(info_struct->ai_family) {
+           case AF_INET: {
+            char address[INET_ADDRSTRLEN];
+            auto *sockaddr = (struct sockaddr_in *)info_struct->ai_addr;
+            inet_ntop(AF_INET, &(sockaddr->sin_addr), address, INET_ADDRSTRLEN);
+            return new StringObject { address };
+          }
+          case AF_INET6: {
+            char address[INET6_ADDRSTRLEN];
+            auto *sockaddr = (struct sockaddr_in6 *)info_struct->ai_addr;
+            inet_ntop(AF_INET6, &(sockaddr->sin6_addr), address, INET6_ADDRSTRLEN);
+            return new StringObject { address };
+          }
+      }
```

Comment:
Should we cache those values that we retrieve from the addrinfo object? They cannot change in the object's lifetime, can they?

Reply:
I don't think it's very costly to look them up, but maybe I'm wrong. We can add caching later if we find there's a bottleneck.